File: python/graphs/a_star_production.py
import heapq
from collections.abc import Iterable


Cell = tuple[int, int]


def a_star_path(
    width: int,
    height: int,
    walls: Iterable[Cell],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    if width <= 0 or height <= 0:
        raise ValueError("grid dimensions must be positive")

    def cell_id(cell: Cell) -> int:
        return cell[0] * width + cell[1]

    def inside(cell: Cell) -> bool:
        row, col = cell
        return 0 <= row < height and 0 <= col < width

    if not inside(start) or not inside(target):
        raise ValueError("start and target must be inside the grid")

    total = width * height
    blocked = bytearray(total)
    for wall in walls:
        if inside(wall):
            blocked[cell_id(wall)] = 1

    start_id = cell_id(start)
    target_id = cell_id(target)
    if blocked[start_id] or blocked[target_id]:
        return []

    def heuristic(cell_id_value: int) -> int:
        row, col = divmod(cell_id_value, width)
        target_row, target_col = target
        return abs(row - target_row) + abs(col - target_col)

    infinity = total + 1
    distance = [infinity] * total
    parent = [-1] * total
    closed = bytearray(total)
    open_set = [(heuristic(start_id), 0, start_id)]
    distance[start_id] = 0

    while open_set:
        _, current_distance, current_id = heapq.heappop(open_set)
        if closed[current_id]:
            continue
        closed[current_id] = 1

        if current_id == target_id:
            path = []
            while current_id != -1:
                path.append(divmod(current_id, width))
                current_id = parent[current_id]
            path.reverse()
            return path

        row, col = divmod(current_id, width)
        for next_row, next_col in (
            (row - 1, col),
            (row, col + 1),
            (row + 1, col),
            (row, col - 1),
        ):
            if not (0 <= next_row < height and 0 <= next_col < width):
                continue

            next_id = next_row * width + next_col
            if blocked[next_id] or closed[next_id]:
                continue

            next_distance = current_distance + 1
            if next_distance >= distance[next_id]:
                continue

            distance[next_id] = next_distance
            parent[next_id] = current_id
            heapq.heappush(open_set, (next_distance + heuristic(next_id), next_distance, next_id))

    return []

File: cpp/graphs/a_star_production.cpp
#include <algorithm>
#include <cstdlib>
#include <limits>
#include <queue>
#include <stdexcept>
#include <utility>
#include <vector>

using Cell = std::pair<int, int>;

struct Candidate {
    int f;
    int g;
    int id;

    bool operator>(const Candidate& other) const {
        return f > other.f;
    }
};

std::vector<Cell> aStarPath(
    int width,
    int height,
    const std::vector<std::vector<bool>>& walls,
    Cell start,
    Cell target
) {
    if (width <= 0 || height <= 0) {
        throw std::invalid_argument("grid dimensions must be positive");
    }

    auto inside = [width, height](Cell cell) {
        return cell.first >= 0 && cell.first < height
            && cell.second >= 0 && cell.second < width;
    };

    if (!inside(start) || !inside(target)) {
        throw std::out_of_range("start and target must be inside the grid");
    }

    if (walls[start.first][start.second] || walls[target.first][target.second]) {
        return {};
    }

    auto id = [width](Cell cell) {
        return cell.first * width + cell.second;
    };

    auto heuristic = [target, width](int cell_id) {
        int row = cell_id / width;
        int col = cell_id % width;
        return std::abs(row - target.first) + std::abs(col - target.second);
    };

    const int total = width * height;
    const int infinity = std::numeric_limits<int>::max();
    std::vector<int> distance(total, infinity);
    std::vector<int> parent(total, -1);
    std::vector<bool> closed(total, false);
    std::priority_queue<Candidate, std::vector<Candidate>, std::greater<>> open;

    const int start_id = id(start);
    const int target_id = id(target);
    distance[start_id] = 0;
    open.push({heuristic(start_id), 0, start_id});

    while (!open.empty()) {
        Candidate current = open.top();
        open.pop();
        if (closed[current.id]) {
            continue;
        }
        closed[current.id] = true;

        if (current.id == target_id) {
            std::vector<Cell> path;
            for (int step_id = target_id; step_id != -1; step_id = parent[step_id]) {
                path.push_back({step_id / width, step_id % width});
            }
            std::reverse(path.begin(), path.end());
            return path;
        }

        Cell cell{current.id / width, current.id % width};
        const std::vector<Cell> neighbors = {
            {cell.first - 1, cell.second},
            {cell.first, cell.second + 1},
            {cell.first + 1, cell.second},
            {cell.first, cell.second - 1},
        };

        for (Cell next : neighbors) {
            if (!inside(next) || walls[next.first][next.second]) {
                continue;
            }

            int next_id = id(next);
            if (closed[next_id]) {
                continue;
            }

            int next_distance = current.g + 1;
            if (next_distance >= distance[next_id]) {
                continue;
            }

            distance[next_id] = next_distance;
            parent[next_id] = current.id;
            open.push({next_distance + heuristic(next_id), next_distance, next_id});
        }
    }

    return {};
}
